### PR TITLE
gcr driver for replication

### PR DESCRIPTION
**Pull request by bitsf**
Change-Id: I5a6626950d3878bfa9726b332e68bee59159269f
Signed-off-by: Ziming Zhang <zziming@vmware.com>

### CODE REVIEW

**Comment by bitsf** [comment]
Because GCR is using project_id as part of namespace, currently we have to extremely specify it as the prefix of target namespace.

File: src/replication/adapter/googlegcr/adapter.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+			model.TriggerTypeManual,
+			model.TriggerTypeScheduled,
+		},
+	}, nil
+}
+
+// HealthCheck checks health status of a registry
+func (d adapter) HealthCheck() (model.HealthStatus, error) {
+	var err error
+	if d.registry.Credential == nil ||
```
Comment (ywk253100) [inline src/replication/adapter/googlegcr/adapter.go]: Is the credential necessary to create a GCR registry?

Comment (bitsf) [reply src/replication/adapter/googlegcr/adapter.go]: Yes, no matter push-based or pull-based, both need credential.
And I also can't find an extra API that can list repos, except using registry native api /v2/_catalog, which can not search by special namespace

**Comment by bitsf** [comment]
ui enhancement PR https://github.com/goharbor/harbor/pull/8028

File: src/common/utils/registry/registry.go [hunk 2]
```diff
@@ -130,9 +128,18 @@ func (r *Registry) Catalog() ([]string, error) {
 	return repos, nil
 }
 
-// Ping ...
+// Ping checks by Head method
 func (r *Registry) Ping() error {
-	req, err := http.NewRequest(http.MethodHead, buildPingURL(r.Endpoint.String()), nil)
+	return r.ping(http.MethodHead)
```
Comment (steven-zou) [inline src/common/utils/registry/registry.go]: Duplicated changes? ecr adapter also has the same change.

File: src/replication/adapter/googlegcr/adapter.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		(len(d.registry.Credential.AccessKey) == 0 && len(d.registry.Credential.AccessSecret) == 0) {
```
Comment (steven-zou) [inline src/replication/adapter/googlegcr/adapter.go]:
```
len(d.registry.Credential.AccessKey) == 0 && len(d.registry.Credential.AccessSecret) == 0
```
=>
```
len(d.registry.Credential.AccessKey) == 0 || len(d.registry.Credential.AccessSecret) == 0
```
?

File: src/common/utils/registry/registry.go (hunk 2 above)
Comment (bitsf) [reply src/common/utils/registry/registry.go]: Yes, because this is two seperate PR, so I copy this code, in case they can't be merged together

**Comment by steven-zou** [comment]
`-0.3%` down a little much more. @bitsf

**Pull request merged**